<launch>
    <env name="ROSCONSOLE_CONFIG_FILE"
         value="$(find arm_robots)/config/val_rosconsole.conf"/>

    <arg name="sim" default="false"/>
    <arg name="gui" default="true"/>
    <arg name="update_description" default="true"/>
    <arg name="verbose" default="false"/>
    <group ns="static_transform_publishers">
        <node pkg="tf2_ros" type="static_transform_publisher" name="rename_origin" required="true"
              args="0 0 0 0 0 0 robot_root world">
        </node>
    </group>

    <group if="$(arg update_description)">
        <!-- uploads robot description and setups up other parameters -->
        <include file="$(find hdt_michigan_moveit)/launch/planning_context.launch">
            <arg name="load_robot_description" value="true"/>
        </include>
    </group>


    <group if="$(arg sim)">
        <arg name="world_name" default="$(find arm_robots)/world/val_walls_sim.world"/>

        <!-- Start gazebo -->
        <group> <!-- this empty group tag creates a scope for the "remap" command -->
            <!-- because we use ros_control only for the joints of the arms, not the grippers,
                 this is used to rename the joint_states topic it published -->
            <remap from="hdt_michigan/joint_states" to="hdt_michigan/arms_joint_states"/>
            <include file="$(find gazebo_ros)/launch/empty_world.launch">
                <arg name="world_name" value="$(arg world_name)"/>
                <arg name="gui" value="$(arg gui)"/>
                <arg name="paused" value="true"/>
                <arg name="verbose" value="$(arg verbose)"/>
            </include>
        </group>

        <!-- Spawn the robot in gazebo -->
        <node name="urdf_spawner" pkg="gazebo_ros" type="spawn_model" respawn="false" output="screen"
              args="-urdf -model hdt_michigan -param robot_description"/>

        <!-- start the ROS controllers, only used in gazebo. This merges gripper and arm joint states -->
        <include file="$(find hdt_michigan_control)/launch/hdt_michigan_control_gazebo.launch">
            <arg name="robot_namespace" value="hdt_michigan"/>
        </include>

        <!-- Move it-->
        <include ns="hdt_michigan" file="$(find hdt_michigan_moveit)/launch/move_group.launch">
            <arg name="allow_trajectory_execution" value="true"/>
            <arg name="moveit_controller_manager" value="gazebo"/>
        </include>
    </group>
    <group unless="$(arg sim)">
        <arg name="world_name" default="$(find arm_robots)/world/val_walls.world"/>

        <!-- Start gazebo -->
        <include file="$(find gazebo_ros)/launch/empty_world.launch">
            <arg name="world_name" value="$(arg world_name)"/>
            <arg name="gui" value="$(arg gui)"/>
            <arg name="paused" value="false"/>
            <arg name="verbose" value="$(arg verbose)"/>
            <arg name="use_sim_time" value="false"/>
        </include>

        <!-- to let moveit send trajectories with the real robot -->
        <include file="$(find arm_robots)/launch/trajectory_follower.launch">
            <arg name="use_val" value="true"/>
        </include>

        <!-- ROS control for real hdt_michigan, does nothing at the moment -->
        <include file="$(find hdt_michigan_control)/launch/hdt_michigan_control.launch">
            <arg name="robot_namespace" value="hdt_michigan"/>
        </include>

        <!-- Move it-->
        <include ns="hdt_michigan" file="$(find hdt_michigan_moveit)/launch/move_group.launch">
            <arg name="allow_trajectory_execution" value="true"/>
            <arg name="moveit_controller_manager" value="real"/>
        </include>
    </group>
</launch>
